### PR TITLE
[WIP] Refactor i2c

### PR DESCRIPTION
**Pull request by julianduque**
This is a work in progress of the implementation of new i2c methods supported in Johnny Five as specified in https://github.com/rwaldron/johnny-five/wiki/IO-Plugins

- [x] Add tests (@julianduque)
- [x] Add Implementation (@MrYsLab)
- [x] Review of i2c implementation (@fivdi)
- [x] Test with real devices
- [ ] Publish `1.2.0`

*Please do not merge, this is a work in progress*

### CODE REVIEW

**Comment by julianduque** [comment]
I finally will get some devices to test on the real board, I hope to have it tomorrow or tuesday, awesome work @MrYsLab and @fivdi :)

**Comment by fivdi** [comment]
@MrYsLab has everything that should have been commited actually been commited? At the moment there is no method called i2cConfig which means nothing can function correctly. There is an incorrectly named method called [i2cCConfig](https://github.com/julianduque/beaglebone-io/blob/refactor-i2c/lib/beaglebone.js#L400) that's never called.

**Comment by MrYsLab** [comment]
@fivdi Wow, you are correct. I will put a debugger on it tomorrow, since the only way the compass should have worked is if the old version of config was called. I updated johnny-five so that shouldn't have happened, so to say the least I am confused. Also, I am not sure why jshint did not find it.

File: lib/beaglebone.js [hunk 1]
```diff
@@ -365,7 +369,195 @@ BeagleBone.prototype.sendI2CReadRequest = function(address, length, cb) {
...
+    this.i2cWrite(address, register);
+  } else {
+    register = 0;
+  }
+
+  event += register;
+
+  this.once(event, handler);
+
+  function readLoop() {
```
Comment (fivdi) [inline lib/beaglebone.js]: `readLoop` is never called, i.e., the flow of control will never reach a point which will result in it being called. Note that we are in `i2cReadOnce` here so continuous reading isn't needed, only a single read is needed.

File: lib/beaglebone.js [hunk 2]
```diff
@@ -365,7 +369,195 @@ BeagleBone.prototype.sendI2CReadRequest = function(address, length, cb) {
...
+    _i2c.i2cRead(address, bytesToRead, buffer, function(i2cError, i2cNumBytes, buffer) {
+      if (i2cError) {
+        this.emit("error", i2cError);
```
Comment (fivdi) [inline lib/beaglebone.js]: `this` is not bound to the correct object here so calling emit will result in an exception being thrown.

File: lib/beaglebone.js [hunk 3]
```diff
@@ -365,7 +369,195 @@ BeagleBone.prototype.sendI2CReadRequest = function(address, length, cb) {
...
+        return;
+      }
+      this.emit(event, buffer);
```
Comment (fivdi) [inline lib/beaglebone.js]: `this` is not bound to the correct object here so calling emit will result in an exception being thrown.

File: lib/beaglebone.js [hunk 4]
```diff
@@ -365,7 +369,195 @@ BeagleBone.prototype.sendI2CReadRequest = function(address, length, cb) {
...
+
+  this.on(event, readLoop);
+
+  function readLoop() {
+    _i2c.i2cRead(address, bytesToRead, buffer, function(i2cError, i2cNumBytes, buffer) {
+      if (i2cError) {
+        this.emit("error", i2cError);
+        return;
+      }
+      this.emit(event, buffer);
```
Comment (fivdi) [inline lib/beaglebone.js]: `this` is not bound to the correct object here so calling emit will result in an exception being thrown.

File: lib/beaglebone.js [hunk 5]
```diff
@@ -365,7 +369,195 @@ BeagleBone.prototype.sendI2CReadRequest = function(address, length, cb) {
...
+  }
+
+  event += register;
+
+  this.on(event, readLoop);
+
+  function readLoop() {
+    _i2c.i2cRead(address, bytesToRead, buffer, function(i2cError, i2cNumBytes, buffer) {
+      if (i2cError) {
+        this.emit("error", i2cError);
```
Comment (fivdi) [inline lib/beaglebone.js]: `this` is not bound to the correct object here so calling emit will result in an exception being thrown.

File: lib/beaglebone.js [hunk 6]
```diff
@@ -365,7 +369,195 @@ BeagleBone.prototype.sendI2CReadRequest = function(address, length, cb) {
...
+    this.i2cWrite(address, register);
+  } else {
+    register = 0;
+  }
+
+  event += register;
+
+  this.on(event, readLoop);
+
+  function readLoop() {
```
Comment (fivdi) [inline lib/beaglebone.js]: `readLoop` is never called, i.e., the flow of control will never reach a point which will result in it being called.

**Comment by fivdi** [comment]
@MrYsLab there are a few fundamental issues with the code at the moment. I guess that either everything hasn't been commited yet, or we need to go back to the drawing board.

**Comment by MrYsLab** [comment]
@fivdi The reason my tests passed was that when I updated johnny-five, it had not been committed to npm and I was using the older version. I see the problems with the latest J5. I have limited time available this week, but I will work on it as time allows. If you have the time and want to jump on this, please do, just let me know.

**Comment by fivdi** [comment]
Ok, I'll take a look.

**Comment by fivdi** [comment]
@julianduque @MrYsLab Some modifications have been commited and are ready for review/testing :)

For testing purposes ds1621.js was used to read the temperature from a DS1621 temperature sensor (it's in the examples directory.) ds1621.js can be removed if you want as it may cause confusion. I don't actully have any I2C devices that work together with Johnny-Five on the BBB, so I've only tested with ds1621.js. Please test with other devices to see how things go.

@julianduque There's a new private method called _i2cRead. I'm not sue how you want to handle private methods so this can be changed if required.

I posted the following comment at the Johnny-Five gitter:

I think might be a flaw in requirements being made of the IO Plugin continuous i2cRead method. Because the method is asynchronous, the IO Plugin implementation can choose to read from the bus in another thread (using worker threads available in C++ addons). If the bus is being accessed by another thread, the JavaScript event loop thread can't know when the bus is being accessed and therefore can't access the bus. For example, i2cWrite (which is a synchronous method) can't be called from JavaScript if i2cRead was previously called as both threads will potentially access the bus at the same time. Note that Galileo-IO does not have this issue because the asynchronous i2cRead method runs in the JavaScript event loop thread.

This means that there is an issue with the current implementation. It may be necessary to use i2c-bus synchronous methods in all cases to avoid the issue. Due to this issue it's currently not possible to read continuously from two devices at the same time.

**Comment by MrYsLab** [comment]
@fivdi The 8x8 bicolor LED works well, but the HMC5883L compass is giving random readings when the device is sitting on a table and not moving. This was not the case for the original i2c methods.

**Comment by fivdi** [comment]
@MrYsLab Thanks for the feedback :) Are you using beaglebone-io/examples/magnetometer.js for the compass test? The Johnny-Five compass implementation was changed to use the new `i2c...` methods a few days ago. Can you tell me whether the johnny-five/lib/compass.js implementation you're using for testing uses the old `sendI2C...` methods or the new `i2c...` methods? It should work with both old and new methods of course.

**Comment by fivdi** [comment]
@MrYsLab Sorry, another question. Above you say "This was not the case for the original i2c methods". Does this mean that the compass circuit you're currently using works correctly with beaglebone-io v1.1.1?

**Comment by MrYsLab** [comment]
@fivdi Yes, I am using beaglebone-io/examples/magnetometer.js and I am using the latest compass.js with the new i2c methods (I checked to make sure this time ;- ) ). For your second question,  the compass worked with v1.1.1. Same hardware configuration, same device, going through J5. The example code did change, since it changed in j5/eg. So it is possible that something broke in the new j5 stuff.

The 8x8 matrix still uses the old style i2c methods, and it still works as it should.

**Comment by fivdi** [comment]
@MrYsLab Ok, thanks. I have an idea what it might be and will give it a try tonight. It looks like I'll have to order a few more devices too :)

**Comment by fivdi** [comment]
So, the idea I had wasn't the best. I thought there might have been something wrong with reading more than two bytes at the same time as the compass reads six at the same time. I tested by reading all four ADC channel data registers from a TSL2561 luminosity sensor using the following code:

```js
  board.i2cConfig(0);

  // Turn TSL2561 on
  board.i2cWriteReg(0x39, 0x80, 3);

  // Read and print all 4 ADC channel data registers
  board.i2cRead(0x39, 0x9c, 4, function (data) {
    console.log(((data[1] << 8) + data[0]) + ' ' + ((data[3] << 8) + data[2]));
  });
```

Everything works as expected. The raw data printed on the console is the same as the raw data (from a slightly modified version) of [this](https://www.npmjs.com/package/tsl2561) module under the same lighting conditions. So, it looks like I need to get some devices supported by Johnny-Five for testing. @julianduque we'll see how things go when you get your devices.

**Comment by fivdi** [comment]
I've ordered an [ADXL345 - Triple-Axis Accelerometer](http://www.adafruit.com/products/1231) and a [HMC5883L - Triple-Axis Magnetometer](http://www.adafruit.com/product/1746) for further testing. Hopefully it doesn't take too long for them to arrive.

**Comment by fivdi** [comment]
With a [HMC5883L - Triple-Axis Magnetometer](http://www.adafruit.com/product/1746) `beaglebone-io/examples/magnetometer.js` functions correctly with the following install:
```
npm install johnny-five@0.8.36
npm install "git://github.com/julianduque/beaglebone-io.git#refactor-i2c"
```
With the following install there are random readings as @MrYsLab states above:
```
npm install johnny-five@0.8.37
npm install "git://github.com/julianduque/beaglebone-io.git#refactor-i2c"
```
The difference between the two setups is the Johnny-Five version. The same version of BeagleBone-IO is used in both cases.

The code for continuous reading from a HMC5883L in Johnny-Five v0.8.37 is [this](https://github.com/rwaldron/johnny-five/blob/master/lib/compass.js#L61-L67):
```
        this.io.i2cRead(address, this.COMMANDS.READREGISTER, READLENGTH, function(bytes) {

          state.x = int16(bytes[0], bytes[1]);
          state.y = int16(bytes[4], bytes[5]);
          state.z = int16(bytes[2], bytes[3]);

        }.bind(this));
```
In v0.8.36 it's this:
```
  // Read Request Loop
  setInterval(function() {
    // Set pointer to X most signficant byte
    this.io.sendI2CWriteRequest(address, preread);

    // Read from register
    this.io.sendI2CReadRequest(address, bytes, data.bind(this));

    // Emit "headingchange" events whenever the calculated
    // heading accessor changes

    // TODO: handling for 360/0
    if (__.range(last - 1, last + 1).indexOf(Math.floor(this.heading)) === -1) {
      this.emit("headingchange", {
        now: this.heading,
        previous: last
      });
    }

    last = Math.floor(this.heading);

  }.bind(this), delay);

```
The difference between the two implementations is that in v0.8.36 there's a sendI2CWriteRequest and a sendI2CReadRequest in every iteration of the continuous read request loop. In v0.8.37 there's a single i2cWrite and then the continuous read request loop only does an i2cRead (see [here](https://github.com/julianduque/beaglebone-io/blob/refactor-i2c/lib/beaglebone.js#L399-L440)). I guess there are similar issues in Galileo-IO.

I'm fairly busy over the next few days and will take a look at fixing it at the weekend.

**Comment by fivdi** [comment]
A HMC5883L Triple-Axis Magnetometer now functions as expected.

**Comment by fivdi** [comment]
So, an ADXL345 Triple-Axis Accelerometer works too :)

The last piece of the puzzle is making multiple devices work concurrently. I think this will work if each device has it's own bus object but haven't modified the code to try it yet. 

@julianduque If your devices have arrived, the code is now in a state where they should work, as long as you only use one device at a time.

**Comment by fivdi** [comment]
With e404e44b3311a50e212b37b56c950505723f0913 multiple devices work concurrently. Everything should work now.

@julianduque How should we proceed from here?

**Comment by julianduque** [comment]
@fivdi awesome!! we need some basic tests and we will be able to publish this

**Comment by fivdi** [comment]
@julianduque do you mean the tests that are currently causing the buid to fail or are you referring to something else?

**Comment by fivdi** [comment]
@MrYsLab if you have a chance, could you do a quick test with your devices to see if everything is working as expected please?

**Comment by MrYsLab** [comment]
@fivdi Brian - nice job!. I will try out the new code in the next couple of days. Should I be using the latest j5 from GIT? If not, which version should I use?

**Comment by fivdi** [comment]
@MrYsLab ok, excellent. Use the latest release on npm, that's 0.8.38 at the moment, but may be different when you give it a try. The `refactor-i2c` branch of BeagleBone-IO should work with "any" recent version of Johnny-Five as it supports the old `sendI2C` and the new `i2c` methods,

**Comment by MrYsLab** [comment]
Tested with the HMC5883L and Adafruit 8x8 bicolor led matrix and both work perfectly. Great job!

**Comment by fivdi** [comment]
That's good news :) Thanks for the feedback.

**Comment by julianduque** [comment]
I'm super busy this week if any of you can finish the pending tests for new methods feel free to bump version and publish to npm, I'll see if I can find some free time this weekend

**Comment by rwaldron** [comment]
@julianduque @fivdi Any progress here? I came across this: http://stackoverflow.com/questions/31346274/object-beaglebone-has-no-method-i2cconfig-trying-to-run-mpu6050-using-beagl

**Comment by fivdi** [comment]
@rwaldron I was hoping that @julianduque would get around to fixing the tests but this doesn't appear to be the case :). I'll take a look at them within the next few days.

**Comment by julianduque** [comment]
I've been super busy at work and out of the loop, @fivdi if you can take care of it it will be awesome, let me know if you need anything from my side.

**Comment by fivdi** [comment]
@julianduque [bStub](https://github.com/julianduque/beaglebone-io/blob/refactor-i2c/test/beaglebone.js#L7-L24) is the stub for bonescript but it implements [several methods](https://github.com/julianduque/beaglebone-io/blob/refactor-i2c/test/beaglebone.js#L17-L22) that bonescript doesn't implement. Is there a reason for this or is the stub simply incorrectly implemented?

**Comment by julianduque** [comment]
@fivdi is incorrectly implemented specially i2c methods.

**Comment by fivdi** [comment]
I'm working on this at the moment. It's taking a little longer than expected as I haven't fully figured out how to use timers with sinon.

**Comment by fivdi** [comment]
@julianduque is it ok to if there are no tests for sendI2CConfig, sendI2CReadRequest, and sendI2CWriteRequest? They are implemented as follows:
```js
BeagleBone.prototype.sendI2CConfig = BeagleBone.prototype.i2cConfig;
BeagleBone.prototype.sendI2CReadRequest = BeagleBone.prototype.i2cReadOnce;
BeagleBone.prototype.sendI2CWriteRequest = BeagleBone.prototype.i2cWrite;
```
There will of course be tests for i2cConfig, i2cReadOnce, and i2cWrite.

**Comment by julianduque** [comment]
@fivdi it's ok :)

**Comment by fivdi** [comment]
@julianduque there are tests for i2c there now. More could be added but it's not bad for a start. If you have a chance, could you take a look a them and let me know what you think. If someone else wants to take a look at them please do too. Any feedback is welcome :)

There is an issue with the [readWithEvent](https://github.com/julianduque/beaglebone-io/blob/refactor-i2c/test/beaglebone.js#L894-L927) test that I don't understand. I've run this test several hundred times successfully on a BeagleBone Black but it doesn't run successfully on Travis. When run locally on a BeagleBone Black, the 8 expected asserts are detected. On Travis, only 7 asserts are detected. That's why [test.expect(8)](https://github.com/julianduque/beaglebone-io/blob/refactor-i2c/test/beaglebone.js#L902) is currently commented out. [readFromRegisterWithEvent](https://github.com/julianduque/beaglebone-io/blob/refactor-i2c/test/beaglebone.js#L831-L862), which is very similar, runs successfully locally on a BeagleBone Black and on Travis.

Tests with real devices were performed with [two-devices.js](https://gist.github.com/fivdi/df237d9624cd87042702) which is the same program that was used for testing the Raspi-IO i2c implementation.

**Comment by julianduque** [comment]
It looks good to me, only one proposal, move `i2c-bus` to `optionalDependencies` and do a check where i2c-bus is called like:

``` js
var i2c;
try {
  i2c = require('i2c-bus')
} catch (e) {
  i2c = i2cBusStub // taken from test
}
```

what do you think?

**Comment by fivdi** [comment]
Perhaps I'm misunderstanding something, but if `i2c-bus` is moved to `optionalDependencies` and `npm install beaglebone-io` is used to install `beaglebone-io`, then `i2c-bus` will not be installed and therefore will not work.

Given that `npm install beaglebone-io` will typically be used to install `beaglebone-io`, I don't think it's a good idea to make `i2c-bus` an optional dependency.

Also, if the following code is used for the tests, then the real `i2c-bus` will be used during testing rather than the stub and this is not desired.

```js
var i2c;
try {
  i2c = require('i2c-bus')
} catch (e) {
  i2c = i2cBusStub // taken from test
}
```

**Comment by fivdi** [comment]
What would the advantage of the proposed change be?

If `i2c-bus` is moved to `optionalDependencies` and can be successfully installed, then the proposed code modifications would result in the real `i2c-bus` being used for tests. This is not desirable as the stub must be used for tests.

**Comment by fivdi** [comment]
If the goal is to run the tests on a Windows PC or Mac then there would be more work involved. `bonescript` would also have to be an optional dependency. I think that's outside the scope of this PR :)

**Comment by julianduque** [comment]
@fivdi ok, so no problem then, this can land a a new version can be released!!

Good job :D

**Pull request merged**